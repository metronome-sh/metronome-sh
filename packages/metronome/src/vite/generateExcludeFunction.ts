import { parse } from "@typescript-eslint/typescript-estree";
import fs from "fs";
import MagicString from "magic-string";
import { walk } from "./walk";

export function generateExcludeFunction({
  configFile,
}: {
  configFile: string | undefined;
}): string {
  if (!configFile) {
    return `
      export default function metronomeExcludeFunction() {
        console.log('Vite config file was not found. Skipping exclude function.');
        return false;
      }
    `;
  }

  const importsUsed = new Set<string>();

  const code = fs.readFileSync(configFile, "utf8");

  const ast = parse(code, {
    jsx: false,
    loc: true,
    range: true,
    comment: true,
    tokens: true,
  });

  // function extractIdentifiers(node: any) {
  //   if (node.type === "Identifier") {
  //     importsUsed.add(node.name);
  //   } else if (node.type === "MemberExpression" && node.object.type === "Identifier") {
  //     importsUsed.add(node.object.name);
  //   }
  //   Object.keys(node).forEach((key) => {
  //     const value = node[key];
  //     if (typeof value === "object" && value !== null) {
  //       extractIdentifiers(value);
  //     }
  //   });
  // }

  let excludeFnContent = "";
  let functionDefined = false;

  // const walk = (node: any, parent?: any) => {

  //   console.log()

  //   if (node.type === "ImportDeclaration") {
  //     console.log({ node });

  //     node.specifiers.forEach((specifier: any) => {
  //       console.log({ import: specifier.local.name });
  //       if (specifier.local && importsUsed.has(specifier.local.name)) {
  //         const [start, end] = node.range;
  //         excludeFnContent += code.substring(start, end) + "\n";
  //       }
  //     });
  //   } else if (node.type === "CallExpression" && node.callee.name === "metronome") {
  //     node.arguments.forEach((arg: any) => {
  //       if (arg.type === "ObjectExpression") {
  //         arg.properties.forEach((prop: any) => {
  //           if (prop.key.name === "unstable_exclude" && prop.value) {
  //             const [start, end] = prop.value.range;
  //             excludeFnContent += `export default ${code.substring(start, end)};\n`;
  //             extractIdentifiers(prop.value);
  //             functionDefined = true;
  //           }
  //         });
  //       }
  //     });
  //   }
  // };

  walk(ast, (node) => {
    if (node.type === "ImportDeclaration") {
      const [start, end] = node.range;

      console.log(code.substring(start, end));
    }
  });

  // If unstable_exclude is not defined, create a default export function that returns false.
  if (!functionDefined) {
    excludeFnContent += "export default function metronomeExcludeFunction() { return false; };\n";
  }

  return excludeFnContent;
}
