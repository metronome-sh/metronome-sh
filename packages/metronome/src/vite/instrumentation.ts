import { Resource } from "@opentelemetry/resources";
import { SemanticResourceAttributes } from "@opentelemetry/semantic-conventions";
import { WebTracerProvider } from "@opentelemetry/sdk-trace-web";
import {
  BatchSpanProcessor,
  ConsoleSpanExporter,
} from "@opentelemetry/sdk-trace-base";
import { METRONOME_VERSION } from "../constants";

export function startInstrumentation() {
  const resource = Resource.default().merge(
    new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: "metronome",
      [SemanticResourceAttributes.SERVICE_VERSION]: METRONOME_VERSION,
    })
  );

  const provider = new WebTracerProvider({
    resource: resource,
  });

  const exporter = new ConsoleSpanExporter();
  const processor = new BatchSpanProcessor(exporter, { maxQueueSize: 1 });
  provider.addSpanProcessor(processor);

  provider.register();
}
