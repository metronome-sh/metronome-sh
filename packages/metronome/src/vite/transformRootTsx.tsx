import MagicString from "magic-string";
import { parse } from "@typescript-eslint/typescript-estree";
import { walk } from "./walk";

export function transformRootTsx({ code, id }: { code: string; id: string }) {
  const ast = parse(code, {
    jsx: true,
    useJSXTextNode: true,
    loc: true,
    range: true,
  });

  const magicString = new MagicString(code, { filename: id });

  magicString.prepend('import { withMetronome } from "metronome-sh/react";\n');

  walk(ast, (node) => {
    if (node.type === "ExportDefaultDeclaration" && node.declaration) {
      const [start, end] = node.declaration.range;
      magicString.overwrite(start, end, `withMetronome(${code.substring(start, end)});`);
    }
  });

  const rootTsx = magicString.toString();

  return {
    code: rootTsx,
    map: new MagicString(rootTsx).generateMap({
      includeContent: true,
      source: id,
      file: id,
    }),
  };
}
