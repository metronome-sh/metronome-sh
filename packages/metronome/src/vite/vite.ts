export { metronome } from "./plugin";
export { registerMetronome } from "./registerMetronome";
export type { MetronomeConfig } from "../common/types";
