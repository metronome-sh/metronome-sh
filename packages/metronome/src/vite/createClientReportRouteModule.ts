import type { ActionFunction, ActionFunctionArgs } from "@remix-run/server-runtime";
import { getIp } from "../common/getIp";

import { METRONOME_VERSION } from "../common/constants";
import { z } from "zod";
import { ServerRouteModule } from "@remix-run/server-runtime/dist/routeModules";

import { MetronomeInternalConfig, RouteMap } from "../common/types";
import { deobfuscate, getRemixAttributes } from "../common/helpers";
import { ClientErrorSchema, PageviewSchema, WebVitalSchema } from "../common/schemas";
import { type AssetsManifest } from "@remix-run/server-runtime/dist/entry";
import { startInstrumentation, tracer } from "../common/instrumentation/Tracer";
import { SemanticAttributes } from "../common/instrumentation/SemanticAttributes";

export const createClientReportRouteModule = ({
  routeMap,
  assetsManifest,
  config,
}: {
  routeMap: RouteMap;
  assetsManifest: AssetsManifest;
  config: MetronomeInternalConfig;
}): ServerRouteModule => {
  const action: ActionFunction = async ({ request }) => {
    startInstrumentation(config);

    const events = await deobfuscate(await request.text());

    const ip = getIp(request) ?? "0.0.0.0";

    if (!events) {
      console.warn("Metronome: couldn't process events");
      return new Response("", { status: 204 });
    }

    const result = z.array(z.any()).safeParse(events);

    if (!result.success) {
      console.warn("Metronome: Invalid event(s)", JSON.stringify(events));
      return new Response("", { status: 204 });
    }

    result.data.forEach((incoming) => {
      const remixAttributes = getRemixAttributes({
        routeMap,
        version: assetsManifest.version,
        path: incoming.pathname,
      });

      // Web vitals
      const wvResult = WebVitalSchema.safeParse(incoming);
      if (wvResult.success) {
        const { data } = wvResult;
        const metric = tracer().createHistogram(data.metric.name);

        metric
          .record(data.metric.value, {
            [SemanticAttributes.WebVitalName]: data.metric.name,
            [SemanticAttributes.WebVitalRating]: data.metric.rating,
            [SemanticAttributes.WebVitalNavigationType]: data.metric.navigationType,
            [SemanticAttributes.MetronomeVersion]: METRONOME_VERSION,
            [SemanticAttributes.ClientAddress]: ip,
            [SemanticAttributes.UserAgentOriginal]: request.headers.get("user-agent") ?? "",
            [SemanticAttributes.HttpPathname]: data.pathname,
            [SemanticAttributes.UrlFull]: data.url,
            [SemanticAttributes.AppHostname]: data.hostname,
            [SemanticAttributes.ClientReferrer]: data.referrer,
            [SemanticAttributes.ClientScreen]: data.screen,
            [SemanticAttributes.ClientLanguage]: data.language,
            [SemanticAttributes.ClientConnection]: data.connection,
            [SemanticAttributes.ClientDeviceCategory]: data.deviceCategory,
            ...remixAttributes,
            ...config.remixPackages,
          })
          .dispose();

        return;
      }

      // Pageviews
      const pvResult = PageviewSchema.safeParse(incoming);

      if (pvResult.success) {
        const metric = tracer().createCounter("pageview");

        const url = new URL(pvResult.data.url);

        metric
          .add(1, {
            [SemanticAttributes.MetronomeVersion]: METRONOME_VERSION,
            [SemanticAttributes.ClientAddress]: ip,
            [SemanticAttributes.UserAgentOriginal]: request.headers.get("user-agent") ?? "",
            [SemanticAttributes.UrlQuery]: url.search,
            [SemanticAttributes.ClientScreen]: pvResult.data.screen,
            [SemanticAttributes.ClientReferrer]: pvResult.data.referrer,
            [SemanticAttributes.AppHostname]: url.hostname,
            [SemanticAttributes.ClientLanguage]: pvResult.data.language,
            [SemanticAttributes.ClientConnection]: pvResult.data.connection,
            [SemanticAttributes.ClientDeviceCategory]: pvResult.data.deviceCategory,
            [SemanticAttributes.HttpPathname]: pvResult.data.pathname,
            ...remixAttributes,
            ...config.remixPackages,
          })
          .dispose();

        return;
      }

      // Client errors
      const ceResult = ClientErrorSchema.safeParse(incoming);

      if (ceResult.success) {
        const { data } = ceResult;

        const span = tracer().startSpan("client_error", {
          attributes: {
            [SemanticAttributes.MetronomeVersion]: METRONOME_VERSION,
            [SemanticAttributes.ClientAddress]: ip,
            [SemanticAttributes.UserAgentOriginal]: request.headers.get("user-agent") ?? "",
            ...remixAttributes,
            ...config.remixPackages,
          },
        });

        span.addEvent(
          "exception",
          {
            [SemanticAttributes.ExceptionType]: data.error.error || "Error",
            [SemanticAttributes.ExceptionEscaped]: false,
            [SemanticAttributes.ExceptionStacktrace]: data.error.stack,
            [SemanticAttributes.ExceptionMessage]: data.error.message,
            [SemanticAttributes.ExceptionFilename]: data.error.filename,
            [SemanticAttributes.ExceptionLineno]: data.error.lineno,
            [SemanticAttributes.ExceptionColno]: data.error.colno,
          },
          data.timestamp
        );

        span.end();

        return;
      }
    });

    return new Response("", { status: 204 });
  };

  return {
    action,
    // action: async (args: ActionFunctionArgs) => {
    //   try {
    //     return await action(args);
    //   } catch (error) {
    //     console.error("Metronome: failed to process events", error);
    //     return new Response("", { status: 204 });
    //   }
    // },
    default: undefined,
  };
};
