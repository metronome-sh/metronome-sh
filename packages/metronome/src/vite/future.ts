import { HeaderAllowlist } from "../common/types";
import { isDeferredData, isResponse } from "./helpers";
import { Span } from "../common/instrumentation/Span";

export function setHeadersToSpan({
  result,
  span,
  headersAllowlist,
}: {
  result: any;
  span: Span;
  headersAllowlist?: HeaderAllowlist;
}) {
  if (!headersAllowlist || (Array.isArray(headersAllowlist) && headersAllowlist.length === 0)) {
    return;
  }

  const type = result instanceof Request ? "request" : "response";

  const headerEntries =
    type === "request"
      ? result.headers
      : isResponse(result)
      ? result.headers
      : isDeferredData(result)
      ? result.init?.headers
      : [];

  transformHeadersInit(headerEntries ?? []).forEach(([key, value]) => {
    if (Array.isArray(headersAllowlist) && !headersAllowlist.includes(key.toLowerCase())) {
      return;
    }
    span.setAttribute(`http.${type}.header.${key}`, value.split(","));
  });
}

function transformHeadersInit(headersInit: HeadersInit): [string, string][] {
  if (Array.isArray(headersInit)) {
    // Already in the desired format
    return headersInit;
  } else if (headersInit instanceof Headers) {
    // Convert Headers object to an array of key-value pairs
    return Array.from(headersInit.entries());
  } else {
    // Convert Record<string, string> to an array of key-value pairs
    return Object.entries(headersInit);
  }
}

function setTraceIdToResponse(response: Response | TypedDeferredData<any>, span: Span) {
  // TODO enable this
  const __metronome = JSON.stringify({
    documentTraceId: span.getContext().traceId,
  });
  const metronomeCookie = `__metronome=${__metronome}; Path=/`;

  if (isResponse(response)) {
    const existing = response.headers.get("Set-Cookie");

    if (existing) {
      response.headers.set("Set-Cookie", `${existing}, ${metronomeCookie}`);
    } else {
      response.headers.set("Set-Cookie", metronomeCookie);
    }

    return;
  }

  if (isDeferredData(response)) {
    if (!response.init) {
      response.init = { headers: { "Set-Cookie": metronomeCookie } };
      return;
    }

    const headers = response.init.headers;

    if (!headers) {
      response.init.headers = { "Set-Cookie": metronomeCookie };
      return;
    }

    if (Array.isArray(headers)) {
      // Try to check if the header already exists
      const existing = headers.find(([key]) => key === "Set-Cookie");

      if (existing) {
        existing[1] = `${existing[1]}, ${metronomeCookie}`;
      } else {
        headers.push(["Set-Cookie", metronomeCookie]);
      }

      return;
    }

    if (headers instanceof Headers) {
      const existing = headers.get("Set-Cookie");

      if (existing) {
        headers.set("Set-Cookie", `${existing}, ${metronomeCookie}`);
      } else {
        headers.set("Set-Cookie", metronomeCookie);
      }

      return;
    }

    const existing = headers["Set-Cookie"];

    if (existing) {
      headers["Set-Cookie"] = `${existing}, ${metronomeCookie}`;
    } else {
      headers["Set-Cookie"] = metronomeCookie;
    }
  }
}
