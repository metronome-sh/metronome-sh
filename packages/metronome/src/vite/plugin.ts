import { Plugin, PluginOption } from "vite";
import MagicString from "magic-string";
import { MetronomeConfig, MetronomeInternalConfig } from "../common/types";
import fs from "fs";
import path from "path";

let metronomeInternalConfig: MetronomeInternalConfig;

export const metronome: (metronomeConfig?: MetronomeConfig) => Plugin = (metronomeConfig) => {
  return {
    apply: "build",
    name: "metronome",
    generateBundle(_, bundle) {
      const serverBuild = Object.entries(bundle).find(([, value]) => {
        const facadeModuleId = (value as any)?.facadeModuleId;
        return facadeModuleId?.includes("virtual:remix/server-build");
      });

      if (!serverBuild) return;

      const [name] = serverBuild;

      const file = bundle[name] as any;

      const magicString = new MagicString(file.code);

      magicString.prepend(`import { registerMetronome } from "metronome-sh/vite";\n`);

      const regex = /const routes = \{([\s\S]*?)\};/m;

      const metronome = JSON.stringify(metronomeInternalConfig, null, 2);

      magicString.replace(regex, (match, p1) => {
        return `export const metronome = ${metronome};\nconst routes = registerMetronome({${p1}}, { version: serverManifest['version'] }, metronome);`;
      });

      file.code = magicString.toString();
      file.map = magicString.generateMap({ hires: true });
    },
    configResolved(config) {
      if (metronomeInternalConfig) return;

      const { root } = config as any;
      const packageJsonPath = path.resolve(root, "package.json");
      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));

      const remixPackages = Object.fromEntries(
        Object.entries(packageJson.dependencies ?? {})
          .filter(([key]) => key.includes("@remix-run/"))
          .map(([key, value]) => [`remix.package.${key.split("/")[1]}`, value])
      ) as Record<string, string>;

      metronomeInternalConfig = {
        ...metronomeConfig,
        remixPackages,
        endpoint: metronomeConfig?.endpoint ?? "https://metrics.metronome.sh",
      };
    },
    transform(code, id) {
      if (id.match(/root\.tsx$/)) {
        const magicString = new MagicString(code);
        magicString.prepend('import { withMetronome } from "metronome-sh/react";\n');
        const defaultExportRegex = /export\sdefault\sfunction/;

        magicString.replace(defaultExportRegex, "const __defaultExport = function");

        magicString.append("export default withMetronome(__defaultExport);");

        return {
          code: magicString.toString(),
          sourcemap: magicString.generateMap({ hires: true }),
        };
      }
    },
  };
};
