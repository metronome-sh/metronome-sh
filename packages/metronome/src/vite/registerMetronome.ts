import { MetronomeInternalConfig, RouteMap, Routes } from "../common/types";
import { wrapRemixFunction } from "./wrapRemixFunction";
import { createClientReportRouteModule } from "./createClientReportRouteModule";
import { type AssetsManifest } from "@remix-run/server-runtime/dist/entry";

export function registerMetronome(
  routes: Routes,
  assetsManifest: Pick<AssetsManifest, "version">,
  config: MetronomeInternalConfig
): Routes {
  if (!config.apiKey) {
    config.apiKey = process.env.METRONOME_API_KEY;
  }

  const routeMap: RouteMap = {};

  for (const [routeId, route] of Object.entries(routes)) {
    routeMap[routeId] = {
      id: routeId,
      parentId: route.parentId,
      path: route.path,
    };

    const newRoute = { ...route, module: { ...route.module } };

    const wrapperOptions = {
      routeId,
      routePath: route.path,
      config,
      assetsManifest,
    };

    if (route.module.action) {
      newRoute.module.action = wrapRemixFunction(route.module.action, {
        type: "action",
        ...wrapperOptions,
      });
    }
    if (route.module.loader) {
      newRoute.module.loader = wrapRemixFunction(route.module.loader, {
        type: "loader",
        ...wrapperOptions,
      });
    }
    routes[routeId] = newRoute;
  }

  // Register custom metronome route
  const baseUrl = "__metronome";

  routes[baseUrl] = {
    id: baseUrl,
    parentId: undefined,
    path: baseUrl,
    index: undefined,
    caseSensitive: undefined,
    module: createClientReportRouteModule({
      routeMap,
      config,
      assetsManifest,
    }),
  };

  return routes;
}
