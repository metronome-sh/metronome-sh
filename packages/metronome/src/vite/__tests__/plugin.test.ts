import { describe, expect, test } from "vitest";
import { $ } from "execa";
import fs from "fs/promises";
import path from "path";

const tmpDir = `${__dirname}/.remix`;
const $$ = $({ cwd: tmpDir });

const viteConfig = `
import { unstable_vitePlugin as remix } from "@remix-run/dev";
import { defineConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import { metronome } from "metronome-sh/vite";

export default defineConfig({
  plugins: [remix(), tsconfigPaths(), metronome()],
});
`.trim();

async function install({ template, dirname }: { template: string; dirname: string }) {
  // For some reason npx has a weird interaction with create-remix in setting up the dirname
  await $$`pnpm dlx create-remix@latest --template ${template} --no-install --yes ${dirname}`;

  const packageJsonPath = path.join(tmpDir, dirname, "package.json");
  const packageJson = JSON.parse(await fs.readFile(packageJsonPath, "utf-8"));

  packageJson.dependencies["metronome-sh"] = "file:../../../../../.";
  await fs.writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2));

  await $$`npm install --prefix ${dirname}`;

  await fs.writeFile(path.join(tmpDir, dirname, "vite.config.ts"), viteConfig);

  await $$`npm run build --prefix ${dirname}`;
}

function cleanup(dirname: string) {
  return $$`rm -rf ${dirname}`;
}

describe(
  "vite plugin",
  () => {
    test("Installs and builds in the express vite template", async () => {
      // Generate random dirname to avoid conflicts
      const dirname = "express-vite-" + Math.random().toString(36).substring(7);

      await install({
        template: "remix-run/remix/templates/unstable-vite-express",
        dirname,
      });

      const serverBuild = await fs.readFile(
        path.join(tmpDir, dirname, "build/server", "index.js"),
        "utf-8"
      );

      // prettier-ignore
      const metronomeImportRegex = /import\s+\{\s*registerMetronome\s*\}\s+from\s+"metronome-sh\/vite";/;

      expect(serverBuild.match(metronomeImportRegex)?.[0]).toBeDefined();
      expect(serverBuild.match(metronomeImportRegex)?.[0]).toMatchSnapshot();

      // prettier-ignore
      const metronomeExportRegex = /export\s+const\s+metronome\s*=\s*\{\s*"remixPackages":\s*\{\s*"remix\.package\.express":\s*"\^[^"]+",\s*"remix\.package\.node":\s*"\^[^"]+",\s*"remix\.package\.react":\s*"\^[^"]+"\s*\},\s*"endpoint":\s*"https:\/\/metrics\.metronome\.sh"\s*\};/

      expect(serverBuild.match(metronomeExportRegex)?.[0]).toBeDefined();
      expect(serverBuild.match(metronomeExportRegex)?.[0]).toMatchSnapshot();

      // prettier-ignore
      const metronomeWrapperRegex = /const\s+routes\s*=\s*registerMetronome\(\s*\{[\s\S]*?"root"[\s\S]*?\}\s*,\s*\{\s*version:\s*serverManifest\['version'\]\s*\}\s*,\s*metronome\s*\);/

      expect(serverBuild.match(metronomeWrapperRegex)?.[0]).toBeDefined();
      expect(serverBuild.match(metronomeWrapperRegex)?.[0]).toMatchSnapshot();

      const withMetronomeRegex = /withMetronome/;

      expect(serverBuild.match(withMetronomeRegex)?.[0]).toBeDefined();
      expect(serverBuild.match(withMetronomeRegex)?.[0]).toMatchSnapshot();

      const files = await fs.readdir(path.join(tmpDir, dirname, "build/client/assets"));

      const rootFile = files.find((file) => file.startsWith("root-"));

      expect(rootFile).toBeDefined();

      const rootBuild = await fs.readFile(
        path.join(tmpDir, dirname, "build/client/assets", rootFile!),
        "utf-8"
      );

      const metronomeQueueRegex = /__metronomeQueue/;

      expect(rootBuild.match(metronomeQueueRegex)?.[0]).toBeDefined();
      expect(rootBuild.match(metronomeQueueRegex)?.[0]).toMatchSnapshot();

      await cleanup(dirname);
    });
  },
  { timeout: 60000 }
);
