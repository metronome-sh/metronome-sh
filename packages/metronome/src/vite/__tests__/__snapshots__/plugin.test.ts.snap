// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vite plugin > Installs and builds in the express vite template 1`] = `"import { registerMetronome } from "metronome-sh/server";"`;

exports[`vite plugin > Installs and builds in the express vite template 2`] = `
"const routes = registerMetronome({
  "root": {
    id: "root",
    parentId: void 0,
    path: "",
    index: void 0,
    caseSensitive: void 0,
    module: route0
  },
  "routes/_index": {
    id: "routes/_index",
    parentId: "root",
    path: void 0,
    index: true,
    caseSensitive: void 0,
    module: route1
  }
}, metronome);"
`;

exports[`vite plugin > Installs and builds in the express vite template 3`] = `"withMetronome"`;

exports[`vite plugin > Installs and builds in the express vite template 4`] = `"__metronomeQueue"`;
